### PR TITLE
Make sure we use the correct SSL object when making a callback

### PR DESCRIPTION
**Pull request by mattcaswell**
When processing a callback within libssl that applies to TLS the original
SSL object may have been created for TLS directly, or for QUIC. When making
the callback we must make sure that we use the correct SSL object. In the
case of QUIC we must not use the internal only SSL object.

Fixes https://github.com/openssl/openssl/issues/25788

### CODE REVIEW

**Comment by mattcaswell** [comment]
Currently this PR targets master only. It needs backport to 3.4, 3.3 and 3.2, but there are some minor merge conflicts backporting to those branches. I will wait until the review of this PR is done before opening PRs for the backports.

**Comment by t8m** [comment]
CI is relevant

**Review by t8m — COMMENTED** [review]
Apart from the CI failure it looks good.

**Comment by mattcaswell** [comment]
Fixups pushed to address the CI failure and various of the style check failures. Some of the style check failures looked spurious though, so I have also added "style: waived".

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
Ping @openssl/committers for second review.

**Review by vdukhovni — COMMENTED** [review]
LGTM. The patch looks correct, but is difficult to be sure it is *complete*, I hope @mattcaswell is sure that all the relevant callback sites were found.

It would have been great if the "inner" SSL handle that might have been used instead, were a different type that the compiler would detect as an inappropriate pointer to pass, and that functions that want the inner SSL handle were suitably modified to expect that. But that's perhaps a future task, if keeping this part of the interface correct proves fragile...

**Review by vdukhovni — APPROVED** [review]

**Comment by mattcaswell** [comment]
I examined every location in the code where we currently call `SSL_CONNECTION_GET_SSL()` to determine which SSL was meant in any given context - and then adjusted accordingly. I hope that this should be sufficient to find everything.

**Comment by mattcaswell** [comment]
Question: It doesn't affect this PR but I wonder what happens in the feature/quic-server branch with callbacks. Have we always been guaranteed to have called `SSL_accept_connection` before the first TLS callback might be called? If not there might not *be* a user SSL to pass to the callback yet!!

**Comment by t8m** [comment]
> Question: It doesn't affect this PR but I wonder what happens in the feature/quic-server branch with callbacks. Have we always been guaranteed to have called `SSL_accept_connection` before the first TLS callback might be called? If not there might not _be_ a user SSL to pass to the callback yet!!

Hmm... this might be unfortunately the case for all the callbacks that can be triggered by the initial handshake frames from the client - perhaps we should delay processing these until SSL_accept_connection() is called?

**Comment by vdukhovni** [comment]
> Question: It doesn't affect this PR but I wonder what happens in the feature/quic-server branch with callbacks. Have we always been guaranteed to have called `SSL_accept_connection` before the first TLS callback might be called? If not there might not _be_ a user SSL to pass to the callback yet!!

Well, the whole point about "The user SSL" is that it is some object to which the user might have attached some "extdata", i.e. something created well before the handshake. In the case of server callbacks, that'd be the acceptor SSL handle that was listening for incoming connections, not the *accepted* SSL that materialises after.

Which handle will be used in callbacks here?

**Comment by vdukhovni** [comment]
> > Question: It doesn't affect this PR but I wonder what happens in the feature/quic-server branch with callbacks. Have we always been guaranteed to have called `SSL_accept_connection` before the first TLS callback might be called? If not there might not _be_ a user SSL to pass to the callback yet!!
> 
> Well, the whole point about "The user SSL" is that it is some object to which the user might have attached some "extdata", i.e. something created well before the handshake. In the case of server callbacks, that'd be the acceptor SSL handle that was listening for incoming connections, not the _accepted_ SSL that materialises after.
> 
> Which handle will be used in callbacks here?

If perhaps some callbacks really need the accepted connection, while others (typically at least those related to session management) need the preëxisting acceptor connection, we have a problem, and potentially need to either copy extdata when creating the accepted SSL handle, or to give the user an early opportunity to do decide how the accepted SSL should be "decorated" (new callback that gets to see both handles), before any other callbacks are invoked.

**Comment by t8m** [comment]
I am going to create a new issue for this.

**Comment by t8m** [comment]
https://github.com/openssl/project/issues/918 - created for the QUIC server handling of callbacks.

**Comment by t8m** [comment]
Merged to the master branch. Thank you @mattcaswell. Please create the backport PR.

**Pull request closed without merging**
